File: creador2d-plugins/unity/Creador2DParallax.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace Creador2D
{
    /// <summary>
    /// Monta las capas de fondo del mundo y las desplaza con la camara.
    ///
    /// Cada capa se instancia como una fila de sprites repetidos: se crean los
    /// necesarios para cubrir el ancho de la camara mas uno a cada lado, y se
    /// reciclan por modulo al desplazarse. Asi el fondo es infinito sin crear
    /// objetos nuevos en tiempo de ejecucion.
    ///
    /// El factor `speedX` es lo que produce la profundidad: 0 deja la capa
    /// clavada a la pantalla y 1 la ancla al mundo. Los valores intermedios
    /// (0,05 para el cielo, 0,15 para el fondo lejano, 0,4 para el medio) son
    /// los que dan la sensacion de distancia.
    /// </summary>
    [ExecuteAlways]
    public class Creador2DParallax : MonoBehaviour
    {
        [Tooltip("Camara de referencia. Si se deja vacio se usa la principal.")]
        public Camera targetCamera;

        [Tooltip("Unidades de Unity por pixel de la imagen de fondo.")]
        public float unitsPerPixel = 1f / 32f;

        [Tooltip("Orden de dibujado base; las capas se colocan por debajo del mundo.")]
        public int baseSortingOrder = -30000;

        class Layer
        {
            public ParallaxInfo Info;
            public Transform Root;
            public List<Transform> Tiles = new List<Transform>();
            public float TileWidth;
        }

        readonly List<Layer> _layers = new List<Layer>();
        Transform _root;

        /// <summary>Crea las capas a partir del export. Sustituye a las anteriores.</summary>
        public void Build(ParallaxInfo[] parallax, float pixelsPerUnitScale)
        {
            Clear();

            if (parallax == null || parallax.Length == 0)
            {
                return;
            }

            unitsPerPixel = pixelsPerUnitScale;

            if (targetCamera == null)
            {
                targetCamera = Camera.main;
            }

            _root = new GameObject("Parallax").transform;
            _root.SetParent(transform, false);

            // Orden de dibujado: lo mas lejano primero.
            Array.Sort(parallax, (a, b) => KindOrder(a.kind).CompareTo(KindOrder(b.kind)));

            for (int i = 0; i < parallax.Length; i++)
            {
                ParallaxInfo info = parallax[i];
                if (!info.visible || string.IsNullOrEmpty(info.imageUrl))
                {
                    continue;
                }

                Texture2D texture = DecodeDataUrl(info.imageUrl);
                if (texture == null)
                {
                    Debug.LogWarning($"[Creador2D] No se pudo decodificar la capa \"{info.name}\".");
                    continue;
                }

                _layers.Add(BuildLayer(info, texture, i));
            }

            LateUpdate();
        }

        Layer BuildLayer(ParallaxInfo info, Texture2D texture, int index)
        {
            var layer = new Layer { Info = info };

            layer.Root = new GameObject($"{info.kind}_{info.name}").transform;
            layer.Root.SetParent(_root, false);

            var sprite = Sprite.Create(
                texture,
                new Rect(0, 0, texture.width, texture.height),
                new Vector2(0.5f, 0.5f),
                1f / unitsPerPixel);

            layer.TileWidth = texture.width * unitsPerPixel;

            // Cuantas copias hacen falta para cubrir la camara con margen.
            float viewWidth = targetCamera != null
                ? targetCamera.orthographicSize * 2f * targetCamera.aspect
                : 40f;
            int copies = Mathf.Max(3, Mathf.CeilToInt(viewWidth / Mathf.Max(0.01f, layer.TileWidth)) + 2);

            if (!info.repeatX)
            {
                copies = 1;
            }

            Color tint = Creador2DGrid.ParseHexColor(info.tint);
            tint.a = Mathf.Clamp01(info.opacity);

            for (int c = 0; c < copies; c++)
            {
                var go = new GameObject($"tile_{c}");
                go.transform.SetParent(layer.Root, false);

                var renderer = go.AddComponent<SpriteRenderer>();
                renderer.sprite = sprite;
                renderer.color = tint;
                // Cada capa va en su propia franja de orden, siempre detras del
                // mundo, que empieza en LayerBaseOrder("GROUND") = -20000.
                renderer.sortingOrder = baseSortingOrder + index * 10;

                layer.Tiles.Add(go.transform);
            }

            return layer;
        }

        void LateUpdate()
        {
            if (targetCamera == null)
            {
                targetCamera = Camera.main;
                if (targetCamera == null)
                {
                    return;
                }
            }

            Vector3 camPos = targetCamera.transform.position;

            foreach (Layer layer in _layers)
            {
                // La capa sigue a la camara solo en la fraccion indicada: la
                // diferencia entre ambas es lo que se percibe como distancia.
                float x = camPos.x * (1f - layer.Info.speedX);
                float y = camPos.y * (1f - layer.Info.speedY) - layer.Info.offsetY * unitsPerPixel;

                if (layer.TileWidth <= 0.01f || layer.Tiles.Count == 1)
                {
                    layer.Tiles[0].position = new Vector3(x, y, 0f);
                    continue;
                }

                // Ancla al multiplo de tile mas cercano para que el reciclado no
                // deje huecos al cruzar de una copia a la siguiente.
                float anchor = Mathf.Floor((camPos.x - x) / layer.TileWidth) * layer.TileWidth;

                for (int i = 0; i < layer.Tiles.Count; i++)
                {
                    float offset = anchor + (i - layer.Tiles.Count / 2) * layer.TileWidth;
                    layer.Tiles[i].position = new Vector3(x + offset, y, 0f);
                }
            }
        }

        public void Clear()
        {
            _layers.Clear();

            if (_root != null)
            {
                if (Application.isPlaying)
                {
                    Destroy(_root.gameObject);
                }
                else
                {
                    DestroyImmediate(_root.gameObject);
                }
                _root = null;
            }
        }

        static int KindOrder(string kind)
        {
            switch (kind)
            {
                case "SKY": return 0;
                case "FAR": return 1;
                case "MID": return 2;
                case "NEAR": return 3;
                default: return 2;
            }
        }

        /// <summary>Convierte `data:image/png;base64,...` en una textura.</summary>
        public static Texture2D DecodeDataUrl(string dataUrl)
        {
            if (string.IsNullOrEmpty(dataUrl))
            {
                return null;
            }

            int comma = dataUrl.IndexOf(',');
            string payload = comma >= 0 ? dataUrl.Substring(comma + 1) : dataUrl;

            try
            {
                byte[] bytes = Convert.FromBase64String(payload);
                var texture = new Texture2D(2, 2, TextureFormat.RGBA32, false);

                if (!texture.LoadImage(bytes))
                {
                    return null;
                }

                // Sin repeticion en el borde el muestreo bilineal mezcla la
                // ultima columna con la primera y aparece una linea fina.
                texture.wrapMode = TextureWrapMode.Repeat;
                texture.filterMode = FilterMode.Bilinear;
                return texture;
            }
            catch (FormatException)
            {
                return null;
            }
        }
    }
}

File: creador2d-plugins/unreal/Source/Creador2D/Private/Creador2DParallax.cpp
#include "Creador2DParallax.h"

#include "Dom/JsonObject.h"
#include "Engine/StaticMesh.h"
#include "Engine/World.h"
#include "ImageUtils.h"
#include "Materials/MaterialInstanceDynamic.h"
#include "Materials/MaterialInterface.h"
#include "Misc/Base64.h"

DEFINE_LOG_CATEGORY_STATIC(LogCreador2DParallax, Log, All);

ACreador2DParallax::ACreador2DParallax()
{
    PrimaryActorTick.bCanEverTick = true;
    Root = CreateDefaultSubobject<USceneComponent>(TEXT("Root"));
    RootComponent = Root;
}

void ACreador2DParallax::BeginPlay()
{
    Super::BeginPlay();
}

int32 ACreador2DParallax::KindOrder(const FString& Kind)
{
    if (Kind == TEXT("SKY")) return 0;
    if (Kind == TEXT("FAR")) return 1;
    if (Kind == TEXT("MID")) return 2;
    if (Kind == TEXT("NEAR")) return 3;
    return 2;
}

void ACreador2DParallax::Build(const TArray<TSharedPtr<FJsonValue>>& ParallaxJson, float InUnitsPerPixel)
{
    ClearLayers();
    UnitsPerPixel = InUnitsPerPixel;

    for (const TSharedPtr<FJsonValue>& Value : ParallaxJson)
    {
        const TSharedPtr<FJsonObject> Object = Value->AsObject();
        if (!Object.IsValid() || !Object->GetBoolField(TEXT("visible")))
        {
            continue;
        }

        FString DataUrl;
        if (!Object->TryGetStringField(TEXT("imageUrl"), DataUrl) || DataUrl.IsEmpty())
        {
            continue;
        }

        FCreador2DParallaxLayer Layer;
        Layer.Kind = Object->GetStringField(TEXT("kind"));
        Layer.Name = Object->GetStringField(TEXT("name"));
        Object->TryGetNumberField(TEXT("speedX"), Layer.SpeedX);
        Object->TryGetNumberField(TEXT("speedY"), Layer.SpeedY);
        Object->TryGetNumberField(TEXT("opacity"), Layer.Opacity);
        Object->TryGetBoolField(TEXT("repeatX"), Layer.bRepeatX);
        Object->TryGetNumberField(TEXT("offsetY"), Layer.OffsetY);

        FString TintHex;
        if (Object->TryGetStringField(TEXT("tint"), TintHex))
        {
            Layer.Tint = FLinearColor::FromSRGBColor(FColor::FromHex(TintHex));
        }

        Layer.Texture = DecodeDataUrl(DataUrl);
        if (!Layer.Texture)
        {
            UE_LOG(LogCreador2DParallax, Warning, TEXT("No se pudo decodificar la capa \"%s\"."), *Layer.Name);
            continue;
        }

        Layers.Add(Layer);
    }

    // Orden de dibujado: lo mas lejano primero.
    Layers.Sort([](const FCreador2DParallaxLayer& A, const FCreador2DParallaxLayer& B) {
        return KindOrder(A.Kind) < KindOrder(B.Kind);
    });

    for (int32 Index = 0; Index < Layers.Num(); ++Index)
    {
        BuildLayer(Layers[Index], Index);
    }

    UE_LOG(LogCreador2DParallax, Log, TEXT("Parallax montado: %d capa(s)."), Layers.Num());
}

void ACreador2DParallax::BuildLayer(FCreador2DParallaxLayer& Layer, int32 Index)
{
    UStaticMesh* PlaneMesh = LoadObject<UStaticMesh>(nullptr, TEXT("/Engine/BasicShapes/Plane.Plane"));
    UMaterialInterface* BaseMaterial = LoadObject<UMaterialInterface>(
        nullptr, TEXT("/Engine/BasicShapes/BasicShapeMaterial.BasicShapeMaterial"));

    if (!PlaneMesh)
    {
        return;
    }

    const float Width = Layer.Texture->GetSizeX() * UnitsPerPixel;
    const float Height = Layer.Texture->GetSizeY() * UnitsPerPixel;
    Layer.TileWidth = Width;

    // Tres copias bastan para cubrir la pantalla y reciclarse: la del centro
    // mas una a cada lado.
    const int32 Copies = Layer.bRepeatX ? 3 : 1;

    for (int32 c = 0; c < Copies; ++c)
    {
        UStaticMeshComponent* Tile = NewObject<UStaticMeshComponent>(this);
        Tile->SetupAttachment(Root);
        Tile->RegisterComponent();
        Tile->SetStaticMesh(PlaneMesh);
        Tile->SetCollisionEnabled(ECollisionEnabled::NoCollision);
        // El plano del motor mide 100 uu y esta en el plano XY; se gira para
        // quedar vertical, de cara a una camara que mira por el eje Y.
        Tile->SetRelativeRotation(FRotator(90.0f, 0.0f, 0.0f));
        Tile->SetWorldScale3D(FVector(Width / 100.0f, Height / 100.0f, 1.0f));

        if (BaseMaterial)
        {
            UMaterialInstanceDynamic* Dynamic = UMaterialInstanceDynamic::Create(BaseMaterial, this);
            Dynamic->SetTextureParameterValue(TEXT("Texture"), Layer.Texture);
            FLinearColor Colour = Layer.Tint;
            Colour.A = Layer.Opacity;
            Dynamic->SetVectorParameterValue(TEXT("Color"), Colour);
            Tile->SetMaterial(0, Dynamic);
        }

        Layer.Tiles.Add(Tile);
    }
}

void ACreador2DParallax::Tick(float DeltaSeconds)
{
    Super::Tick(DeltaSeconds);

    APlayerController* Controller = GetWorld() ? GetWorld()->GetFirstPlayerController() : nullptr;
    if (!Controller || !Controller->PlayerCameraManager)
    {
        return;
    }

    const FVector CamPos = Controller->PlayerCameraManager->GetCameraLocation();

    for (int32 Index = 0; Index < Layers.Num(); ++Index)
    {
        FCreador2DParallaxLayer& Layer = Layers[Index];
        if (Layer.Tiles.Num() == 0)
        {
            continue;
        }

        // La capa sigue a la camara solo en la fraccion indicada.
        const float X = CamPos.X * (1.0f - Layer.SpeedX);
        const float Z = CamPos.Z * (1.0f - Layer.SpeedY) - Layer.OffsetY * UnitsPerPixel;
        const float DepthY = BaseDepthY + Index * 10.0f;

        if (Layer.Tiles.Num() == 1 || Layer.TileWidth <= 1.0f)
        {
            Layer.Tiles[0]->SetWorldLocation(FVector(X, DepthY, Z));
            continue;
        }

        // Ancla al multiplo de tile mas cercano: sin esto, al reciclar las
        // copias quedarian huecos al cruzar de una a la siguiente.
        const float Anchor = FMath::FloorToFloat((CamPos.X - X) / Layer.TileWidth) * Layer.TileWidth;

        for (int32 t = 0; t < Layer.Tiles.Num(); ++t)
        {
            const float Offset = Anchor + (t - Layer.Tiles.Num() / 2) * Layer.TileWidth;
            Layer.Tiles[t]->SetWorldLocation(FVector(X + Offset, DepthY, Z));
        }
    }
}

void ACreador2DParallax::ClearLayers()
{
    for (FCreador2DParallaxLayer& Layer : Layers)
    {
        for (UStaticMeshComponent* Tile : Layer.Tiles)
        {
            if (IsValid(Tile))
            {
                Tile->DestroyComponent();
            }
        }
    }

    Layers.Empty();
}

UTexture2D* ACreador2DParallax::DecodeDataUrl(const FString& DataUrl)
{
    int32 Comma = INDEX_NONE;
    DataUrl.FindChar(TEXT(','), Comma);

    const FString Payload = Comma != INDEX_NONE ? DataUrl.RightChop(Comma + 1) : DataUrl;

    TArray<uint8> Bytes;
    if (!FBase64::Decode(Payload, Bytes) || Bytes.Num() == 0)
    {
        return nullptr;
    }

    UTexture2D* Texture = FImageUtils::ImportBufferAsTexture2D(Bytes);
    if (Texture)
    {
        // Sin repeticion, el muestreo del borde mezcla la ultima columna con la
        // primera y aparece una linea fina en cada union.
        Texture->AddressX = TA_Wrap;
        Texture->AddressY = TA_Clamp;
        Texture->UpdateResource();
    }

    return Texture;
}
